Create big grid to use for inventorying WQP data

lekoenig opened the pull request:
Addresses #10.

This PR creates a big, CONUS-scale grid that can be used to chunk data queries to WQP. One target (`p1_conus_grid`) creates the grid and another target (`p1_overlapping_boxes`) subsets the big grid to find those boxes that overlap with a buffered version of our AOI polygon. I plan on using dynamic branching to map the WQP data queries over the boxes within `p1_overlapping_boxes`.

I'm open to any feedback you have here, including target names, documentation, etc. I considered moving the code to build `p1_conus_grid` to a separate function to clean up the fetch phase file, so let me know if you have any thoughts on that. Thanks in advance!

Here's what our pipeline looks like including the changes in this PR:

![pipeline_makegrid](https://user-images.githubusercontent.com/8785034/159369781-5066d6ab-9358-4e46-a590-8bc627a78c10.png)

And here's a preview of the output:

![boxes](https://user-images.githubusercontent.com/8785034/159370194-a0a90c36-0f1d-46d7-9be9-40a0a1a08496.png)

lindsayplatt commented on 1_fetch.R:
```diff
@@ -36,6 +36,40 @@ p1_targets_list <- list(
+  # Create a big grid of boxes to set up chunked data queries
+  tar_target(
+    p1_conus_grid,
+    tigris::states(class = "sf", year = 2020, progress_bar = FALSE) %>%
```
I like the idea of this grid code and the code for getting the grid that intersects with the aoi to be inside their own script just to clean things up. I know it seems that I may be a bit all over the place with my feedback about whether to put code in a script or leave here, but I tend to sweep code into a custom function when it is more than 3 lines of code or so. I also do this when I think the code itself is not super understandable to a user and I think I can make a function name that is more helpful/descriptive.

lekoenig replied:
I appreciate your input here, @lindsayplatt! I agree that having the code within a function cleans up the phase file and I like your benchmarks for when to use a separate function. I've been trying to think about what would be most easily understood/interpreted for anyone looking at the repo, even if they don't have a lot of experience with `targets`. I also like the idea of moving the code to build these two targets to separate functions because these code steps are somewhat tangential to the main task of downloading data from WQP. This way a user isn't as likely to get bogged down thinking they need to interpret the code (unless they want to dig into the functions further).

lindsayplatt replied:
That's a great plan for this repo - separate the parts that are tangential to the main idea into their own functions and showcase the code behind the main targets. I like that!

lindsayplatt reviewed (commented):
- I can run this and get the same results 🎉
- One question about projections because I am not following why the new one is necessary
- Leaving suggestions that are optional for naming/code locations

lekoenig commented on 1_fetch.R:
```diff
@@ -36,6 +36,40 @@ p1_targets_list <- list(
+  # (buffered here using a 5 km buffer distance). These boxes will be used
+  # to query the WQP.
+  tar_target(
+    p1_overlapping_boxes,
```
I like this suggestion much better - `p1_conus_grid_aoi` it is!

lindsayplatt commented on 1_fetch.R:
```diff
@@ -36,6 +36,40 @@ p1_targets_list <- list(
+    {
+      # Project area of interest to calculate buffer and set up for intersection
+      buffered_AOI <- p1_AOI_sf %>%
+        sf::st_transform(5070) %>%
```
I'm not as versed in the `st_filter()` methods and appropriate ways to include a buffer, so I will defer to you on this one!

lekoenig commented:
Thanks for your review, @lindsayplatt! I've incorporated your suggestions, namely, I edited the target name for the boxes that overlap our AOI to make it more informative, and moved the code that creates the big grid and subset of boxes from the phase file to separate functions. 

I ended up going with a within-distance approach for subsetting the big grid as opposed to buffering the AOI polygon. The two approaches should function the same way for our use case, but the within-distance approach simplifies the code steps and might even be a tiny bit faster too. The output of `p1_conus_grid_aoi` should look the same as before:

![AOI_2](https://user-images.githubusercontent.com/8785034/159564678-e2907ec1-eade-4869-8456-9a0e3b0929a4.png)

lindsayplatt commented on 1_fetch/src/create_grids.R:
```diff
@@ -0,0 +1,62 @@
+  #' @description Function to create a big grid of boxes to use for chunked
+  #' data queries
+  #' 
+  #' @param cellsize the target cell size of each box in map units (here, degrees);
```
might want to explain that it needs two values - size in x direction and size in y direction

lekoenig replied:
Good point, I've made that change in [a3046a2](https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/15/commits/a3046a240e3132472ed3c187f70d0128d642b375).

lekoenig commented on 1_fetch/src/create_grids.R:
```diff
@@ -0,0 +1,62 @@
+  #' within the holistic grid that overlaps the buffered area of interest.
+  #' 
+  #' @example subset_grids_to_aoi(grid = p1_conus_grid, dist_m = 5000)
+  #' 
```
I don't have a good reason 😄. Do you think documentation on the outside is more common across USGS-R repos? I'm happy to defer to your preference here.

lindsayplatt replied:
I've only ever seen it on the outside (typically right above), so I wasn't sure if there was any specific reason for doing it on the inside :)

lekoenig replied:
Sounds good to me, let's go with documentation outside 😺. I've made that change for the two functions in `1_fetch/src/create_grids.R`.
